File: MiniProject/Teacher.java
package MiniProject;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;

public class Teacher implements Serializable
{
    private final String name;
    private final String teacherId;
    private Set<String> coursesId = new HashSet<>();

    public Teacher(String name, String teacherId, Set<String> coursesId)
    {
        this.name = name;
        this.teacherId = teacherId;
        this.coursesId = coursesId;
    }

    public String getName()
    {
        return name;
    }

    public String getTeacherId()
    {
        return teacherId;
    }

    public int getCoursesNumber()
    {
        return coursesId.size();
    }

    public Set<String> getCoursesId()
    {
        return coursesId;
    }

    public void addCourse(Course course)
    {
        coursesId.add(course.getCourseId());
    }

    public void addCourse(String courseId)
    {
        coursesId.add(courseId);
    }

    public void removeCourse(Course course)
    {
        coursesId.remove(course.getCourseId());
    }

    public void removeCourse(String courseId)
    {
        coursesId.remove(courseId);
    }


    @Override
    public boolean equals(Object obj)
    {
        if(!(obj instanceof Teacher))
        {
            return false;
        }
        Teacher teacher = (Teacher)obj;

		return teacherId.equals(teacher.getTeacherId());
    }

    @Override
    public String toString()
    {
        return "Teacher{" + "name='" + name + '\'' + ", teacherId='" + teacherId + '\'' + ", coursesId=" + coursesId
                + '}';
    }
}

File: android_front/lib/classes/Teacher.dart
import 'dart:collection';
import 'Course.dart';

class Teacher {
  final String name;
  final String teacherId;
  Set<String> coursesId;

  Teacher(this.name, this.teacherId, Set<String> coursesId)
      : coursesId = HashSet<String>.from(coursesId);

  String getName() => name;

  String getTeacherId() => teacherId;

  int getCoursesNumber() => coursesId.length;

  Set<String> getCoursesId() => coursesId;

  void addCourse(Course course) {
    coursesId.add(course.courseId);
  }

  void addCourseById(String courseId) {
    coursesId.add(courseId);
  }

  void removeCourse(Course course) {
    coursesId.remove(course.courseId);
  }

  void removeCourseById(String courseId) {
    coursesId.remove(courseId);
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other is! Teacher) return false;
    return teacherId == other.teacherId;
  }

  @override
  int get hashCode => teacherId.hashCode;

  @override
  String toString() {
    return 'Teacher{name=$name, teacherId=$teacherId, coursesId=$coursesId}';
  }
}
